[stdlib] Implement partition API change (SE-0120)

natecook1000 opened the pull request:
<!-- Please complete this template before creating the pull request. -->
#### What's in this pull request?

This implements the partition API changes proposed in [SE-0120](https://github.com/apple/swift-evolution/blob/master/proposals/0120-revise-partition-method.md).

**Please don't merge until/unless the proposal is accepted.**
#### Resolved bug number: ([SR-1965](https://bugs.swift.org/browse/SR-1965))

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>



The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

natecook1000 committed:
[stdlib] Implement partition API change (SE-0120)

natecook1000 committed:
[stdlib] Update tests for new partition APIs

Special thanks to @aschwaighofer for help with these tests and fix-its!

gribozavr commented on stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb:
```diff
@@ -907,16 +907,20 @@ func checkPartition_${'Predicate' if predicate else 'WhereElementIsComparable'}(
   var c = makeWrappedCollectionWithComparableElement(elements)
 %   end
-
+  let pivotElt  = c.first
```
Extra space before `=`.

gribozavr commented on stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb:
```diff
@@ -907,16 +907,20 @@ func checkPartition_${'Predicate' if predicate else 'WhereElementIsComparable'}(
-    _blackHole(closureLifetimeTracker)
-    return extract(lhs).value < extract(rhs).value
+  if let first = pivotElt {
+    pivot = c.partition(by: {
```
Why not try calling `partition` on empty collections?  That's important to test.  Feel free to unconditionally unwrap `pivotElt` in the closure, it should never get executed.

natecook1000 replied:
Thanks @gribozavr! I've modified this test to take a pivot value rather than pivoting on the first element. This way we can also test the case where no elements in the collection pass the predicate and every element in the collection passes. Given that `partition(by:)` doesn't care about the Comparable-ness of the elements, I'm not sure how important the Opaque wrapper is, but I've left it in so we could use the identity-checking mechanism.

gribozavr commented on stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb:
```diff
@@ -907,16 +907,20 @@ func checkPartition_${'Predicate' if predicate else 'WhereElementIsComparable'}(
-  let pivot = c.partition()
+  if let first = c.first {
+    pivot = c.partition(by: { !($0 < first) })
+  }
```
Is there still a `partition()` without a predicate?  If not, just drop this branch of the test.

gribozavr commented on stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb:
```diff
@@ -988,13 +992,18 @@ self.test("\(testNamePrefix).partition/DispatchesThrough_withUnsafeMutableBuffer
+  var pivot = lc.startIndex
+  if let first = lc.first {
+    pivot = lc.partition(by: { !($0 < first) })
+  }
```
Same, if there is no `partition` without the predicate, please drop this branch.

gribozavr commented on stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb:
```diff
@@ -988,13 +992,18 @@ self.test("\(testNamePrefix).partition/DispatchesThrough_withUnsafeMutableBuffer
-    return extract(lhs).value < extract(rhs).value
+  var pivot = lc.startIndex
+  if let first = c.first {
+    pivot = lc.partition(by: {
```
Please run tests on empty collections.

gribozavr commented on test/Interpreter/protocol_extensions.swift:
```diff
@@ -76,7 +76,10 @@ extension MutableCollection
+    if let first = self.first {
+      return self.partition(by: { $0 >= first})
+    }
+    return self.startIndex
```
You can replace the body with just: `return self.partition(by: { $0 >= first! })`

natecook1000 replied:
Replaced, good call

gribozavr commented on validation-test/stdlib/Algorithm.swift:
```diff
@@ -149,7 +149,10 @@ Algorithm.test("invalidOrderings") {
+    let lt = $0
+    if let first = a.first {
+      _ = a.partition(by: { !lt($0, first) })
+    }
```
Same.

gribozavr commented on stdlib/public/core/CollectionAlgorithms.swift.gyb:
```diff
@@ -142,149 +142,84 @@ orderingExplanation = """\
-%   end
-  {
-    let maybeOffset = _withUnsafeMutableBufferPointerIfSupported {
+extension MutableCollection where Self: BidirectionalCollection {
```
Please add a space before the colon.

I'd also be surprised if the existing test compiled, since it probably requires `RandomAccessCollection`.

gribozavr commented on stdlib/public/core/CollectionAlgorithms.swift.gyb:
```diff
@@ -142,149 +142,84 @@ orderingExplanation = """\
   public mutating func partition(
-    isOrderedBefore: @noescape (${IElement}, ${IElement}) -> Bool
-  ) -> Index
+      by belongsInSecondPartition: @noescape (${IElement}) throws -> Bool
```
You'd need a new test for forward collections that would resolve to this overload.

gribozavr replied:
Actually, since you have different implementations now, you should make `partition()` a requirement of `MutableCollection` so that we get the most efficient implementation via dynamic dispatch.  Then you'd need just one test for forward collections, it will dispatch to both implementations depending on the type.

natecook1000 replied:
The requirement is in, so we should be getting dynamic dispatch. Will the existing partition tests run for forward collections, or do I need to make a change to enable that?

gribozavr replied:
You will need to move the test in `CheckMutableCollectionType.swift.gyb` into the `addMutableCollectionTests` function (that only requires forward collections).

natecook1000 replied:
Thanks, I totally didn't get how any of that was organized… The main partition tests are now running on all mutable collections—I kept the `_withUnsafeMutableBufferPointerIfSupported` test in the random-access section for now.

natecook1000 committed:
[stdlib] Update partition implementation with feedback

natecook1000 committed:
[stdlib] Update partition tests with feedback

natecook1000 committed:
Test partition(by:) on all mutable collections

natecook1000 commented:
@swift-ci Please test

gribozavr commented on stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb:
```diff
@@ -575,6 +575,87 @@ self.test("\(testNamePrefix).sorted/${'Predicate' if predicate else 'WhereElemen
+  lessImpl: ((Int, Int) -> Bool),
+  verifyOrder: Bool
+) {
+  let extract = extractValue
```
Can you remove the `extract` variable and just use `extractValue`?

gribozavr commented on stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb:
```diff
@@ -879,123 +960,24 @@ self.test("\(testNamePrefix).sort/${'Predicate' if predicate else 'WhereElementI
-}
-
-self.test("\(testNamePrefix).partition/DispatchesThrough_withUnsafeMutableBufferPointerIfSupported/${'Predicate' if predicate else 'WhereElementIsComparable'}") {
+// FIXME(tests): Move to addMutableCollectionTests?
```
Could you do this?  Or is there something blocking you?

gribozavr commented on stdlib/private/StdlibCollectionUnittest/CheckMutableCollectionType.swift.gyb:
```diff
@@ -879,123 +960,24 @@ self.test("\(testNamePrefix).sort/${'Predicate' if predicate else 'WhereElementI
+self.test("\(testNamePrefix).partition/DispatchesThrough_withUnsafeMutableBufferPointerIfSupported") {
   let sequence = [ 5, 4, 3, 2, 1 ]
-%   if predicate:
   let extract = extractValue
```
Can you remove the `extract` variable and just use `extractValue`?

gribozavr commented on stdlib/public/core/MutableCollection.swift:
```diff
@@ -267,6 +267,46 @@ public protocol MutableCollection : MutableIndexable, Collection {
+  ///   equal to the collection's `endIndex`.
+  ///
+  /// - Complexity: O(n)
+  mutating func partition(
```
Since this is a new requirement, could you add a dispatch test in `validation-test/stdlib/CollectionType.swift.gyb`, next to the test for `_withUnsafeMutableBufferPointerIfSupported`?

natecook1000 replied:
@gribozavr These changes are all in—now testing for dispatch and for unsafe mutable buffer pointer access for bidirectional and random-access collections. Thanks!

natecook1000 committed:
[stdlib] Update partition(by:) tests

Adds a dispatch test for partition(by:), since it's now a protocol
requirement. Also adds a new logging collection wrapper that only
logs when _withUnsafeMutableBufferPointerIfSupported is called --
any calls to this method from dispatched methods are uncounted by
the standard logging wrappers.

natecook1000 commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test OS X platform

gribozavr commented on stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb:
```diff
@@ -566,6 +574,95 @@ public struct ${Self}<
+  public mutating func _withUnsafeMutableBufferPointerIfSupported<R>(
+    _ body: @noescape (UnsafeMutablePointer<Iterator.Element>, Int) throws -> R
+  ) rethrows -> R? {
+    print("Log._withUnsafeMutableBufferPointerIfSupported[selfType] += 1")
```
Debugging print?

natecook1000 replied:
😳 Removed.

natecook1000 committed:
Remove leftover print

gribozavr commented:
@swift-ci Please smoke test and merge
